File: services/memory-service/app/api/routes/sessions.py
"""会话路由：CRUD / 消息历史 / 追加消息（docs/01 §6.3，docs/06 §M-03）。

workspace 归属经 query 参数 workspace_id 传递，由
CurrentWorkspaceQuery 依赖完成成员校验（隔离第二道防线）。
"""
import uuid
from typing import Annotated

from fastapi import APIRouter, Query, status

from app.core.deps import CurrentUser, CurrentWorkspaceQuery, DbDep
from app.core.errors import AppError
from app.models.enums import SessionStatus
from app.schemas.common import Page
from app.schemas.message import MessageRead
from app.schemas.session import (
    MessageCreate,
    SessionCreate,
    SessionDetail,
    SessionRead,
    SessionUpdate,
)
from app.services.session_service import SessionService

router = APIRouter(prefix="/api/sessions", tags=["sessions"])

_service = SessionService()

# 分页 query 参数（列表与消息历史共用约束）
LimitQuery = Annotated[int, Query(ge=1, le=100, description="页大小（1-100）")]
OffsetQuery = Annotated[int, Query(ge=0, description="偏移量")]


@router.get("", response_model=Page[SessionRead], summary="会话列表（分页、关键词）")
async def list_sessions(
    ws: CurrentWorkspaceQuery,
    db: DbDep,
    limit: LimitQuery = 20,
    offset: OffsetQuery = 0,
    q: Annotated[str | None, Query(max_length=100, description="标题关键词")] = None,
    include_archived: Annotated[bool, Query(description="是否含归档会话")] = False,
) -> Page[SessionRead]:
    """当前 workspace 的会话列表（软删除始终排除）。"""
    sessions, total = await _service.list_paginated(
        db,
        ws_id=ws.id,
        limit=limit,
        offset=offset,
        keyword=q,
        include_archived=include_archived,
    )
    return Page(items=sessions, total=total)


@router.post(
    "",
    response_model=SessionRead,
    status_code=status.HTTP_201_CREATED,
    summary="创建会话",
)
async def create_session(
    payload: SessionCreate, ws: CurrentWorkspaceQuery, user: CurrentUser, db: DbDep
) -> SessionRead:
    """创建会话；title 为空时用占位（首条消息后由摘要命名）。"""
    return await _service.create(db, ws_id=ws.id, user=user, title=payload.title)


@router.get("/{session_id}", response_model=SessionDetail, summary="会话详情")
async def get_session(
    session_id: uuid.UUID, ws: CurrentWorkspaceQuery, db: DbDep
) -> SessionDetail:
    """会话详情（含首页消息；完整历史走 messages 子端点）。"""
    return await _service.get_detail(db, ws_id=ws.id, session_id=session_id)


@router.patch("/{session_id}", response_model=SessionRead, summary="重命名 / 归档 / 恢复")
async def update_session(
    session_id: uuid.UUID,
    payload: SessionUpdate,
    ws: CurrentWorkspaceQuery,
    db: DbDep,
) -> SessionRead:
    """按提交字段部分更新（title / status；删除是独立端点）。"""
    if payload.status is SessionStatus.DELETED:
        raise AppError("invalid_operation", 422, "删除请使用 DELETE 端点")
    if payload.title is not None:
        await _service.rename(db, ws_id=ws.id, session_id=session_id, title=payload.title)
    if payload.status is SessionStatus.ARCHIVED:
        await _service.archive(db, ws_id=ws.id, session_id=session_id)
    elif payload.status is SessionStatus.ACTIVE:
        await _service.restore(db, ws_id=ws.id, session_id=session_id)
    refreshed = await _service.get(db, ws_id=ws.id, session_id=session_id)
    return SessionRead.model_validate(refreshed)


@router.delete("/{session_id}", status_code=status.HTTP_204_NO_CONTENT, summary="删除会话")
async def delete_session(
    session_id: uuid.UUID,
    ws: CurrentWorkspaceQuery,
    db: DbDep,
    cascade_memories: Annotated[
        bool, Query(description="true 时同步归档该会话来源的记忆")
    ] = False,
) -> None:
    """软删除；cascade_memories=True 时同步归档来源记忆。"""
    await _service.delete(
        db, ws_id=ws.id, session_id=session_id, cascade_memories=cascade_memories
    )


@router.get(
    "/{session_id}/messages",
    response_model=Page[MessageRead],
    summary="消息历史（时间正序分页）",
)
async def list_messages(
    session_id: uuid.UUID,
    ws: CurrentWorkspaceQuery,
    db: DbDep,
    limit: LimitQuery = 50,
    offset: OffsetQuery = 0,
) -> Page[MessageRead]:
    """会话消息历史（含引用与 trace 链接的 metadata）。"""
    messages, total = await _service.list_messages(
        db, ws_id=ws.id, session_id=session_id, limit=limit, offset=offset
    )
    return Page(items=messages, total=total)


@router.post(
    "/{session_id}/messages",
    response_model=MessageRead,
    status_code=status.HTTP_201_CREATED,
    summary="追加消息（锁 + 乐观锁）",
)
async def append_message(
    session_id: uuid.UUID,
    payload: MessageCreate,
    ws: CurrentWorkspaceQuery,
    db: DbDep,
) -> MessageRead:
    """追加消息：token 计数、metadata 落库、版本冲突返回 409。"""
    return await _service.append_message(
        db, ws_id=ws.id, session_id=session_id, payload=payload
    )

File: frontend/src/api/sessions.ts
/**
 * 会话 API（EchoDesk 前端）。
 *
 * 对接 memory-service `/api/sessions/*`：CRUD + 消息历史分页。
 * 注意：该路由族 workspace_id 一律走 query 参数（服务端 CurrentWorkspaceQuery），
 * body 只承载业务字段。
 */
import { api } from "./client";

/** 会话信息。 */
export interface SessionRead {
  id: string;
  workspace_id: string;
  user_id: string;
  title: string;
  status: string;
  token_total: number;
  last_message_at: string | null;
  version: number;
  created_at: string;
  updated_at: string;
}

/** 会话消息。 */
export interface MessageRead {
  id: string;
  session_id: string;
  role: "user" | "assistant" | "tool" | "system";
  content: string;
  token_count: number;
  version: number;
  metadata: Record<string, unknown>;
  created_at: string;
}

/** 分页容器。 */
export interface Page<T> {
  items: T[];
  total: number;
}

/**
 * 列出当前 workspace 的会话（updated_at 降序）。
 *
 * @param workspaceId - workspace ID。
 * @param limit - 单页条数（服务端上限 100）。
 * @returns 会话分页。
 */
export async function listSessions(workspaceId: string, limit = 50): Promise<Page<SessionRead>> {
  const { data } = await api.get<Page<SessionRead>>("/api/sessions", {
    params: { workspace_id: workspaceId, limit },
  });
  return data;
}

/**
 * 创建会话。
 *
 * @param workspaceId - workspace ID。
 * @param title - 标题（可空，服务端用占位、首条消息后摘要命名）。
 * @returns 新会话。
 */
export async function createSession(workspaceId: string, title?: string): Promise<SessionRead> {
  const { data } = await api.post<SessionRead>("/api/sessions", title === undefined ? {} : { title }, {
    params: { workspace_id: workspaceId },
  });
  return data;
}

/**
 * 更新会话（重命名 / 归档）。
 *
 * @param workspaceId - workspace ID。
 * @param sessionId - 会话 ID。
 * @param patch - 待更新字段。
 * @returns 更新后的会话。
 */
export async function updateSession(
  workspaceId: string,
  sessionId: string,
  patch: { title?: string; status?: string }
): Promise<SessionRead> {
  const { data } = await api.patch<SessionRead>(`/api/sessions/${sessionId}`, patch, {
    params: { workspace_id: workspaceId },
  });
  return data;
}

/**
 * 删除会话（软删除）。
 *
 * @param workspaceId - workspace ID。
 * @param sessionId - 会话 ID。
 */
export async function deleteSession(workspaceId: string, sessionId: string): Promise<void> {
  await api.delete(`/api/sessions/${sessionId}`, { params: { workspace_id: workspaceId } });
}

/**
 * 拉取会话消息历史（时间正序）。
 *
 * @param workspaceId - workspace ID。
 * @param sessionId - 会话 ID。
 * @param limit - 单页条数（服务端上限 100）。
 * @param offset - 偏移量（长会话取最近一页：offset = max(0, total - limit)）。
 * @returns 消息分页。
 */
export async function listMessages(
  workspaceId: string,
  sessionId: string,
  limit = 100,
  offset = 0
): Promise<Page<MessageRead>> {
  const { data } = await api.get<Page<MessageRead>>(`/api/sessions/${sessionId}/messages`, {
    params: { workspace_id: workspaceId, limit, offset },
  });
  return data;
}
